### ISSUE TITLE
Support for writing GeoTIFF files

### ISSUE THREAD

**Issue by gschulze** [georust/geotiff#16]
It would be nice also to be able to write GeoTIFFs. Given that the _tiff_ crate is used in the background, it should be straightforward to implement this. I can offer to implement this.

**Comment by weiji14** [comment]
Hi @gschulze, yes please, go for it! The migration to the `tiff` crate isn't fully complete yet (been caught up in travel lately so haven't found the time), but do take a look around and open PRs as needed.

**Comment by gschulze** [comment]
Hi, @weiji14! I managed to complete the migration to the `tiff` crate, although it might initially seem a bit radical. The corresponding PR is #17.

Next, I would do the coordinate transformation stuff, as, in my opinion, this is what should be provided by this crate on top of the functionality already in the `tiff` crate.

Writing GeoTIFFs should be straightforward, then. What is missing is the functionality to retrieve all tags from a file so that files can be read, modified, and written to disk again. I filed a [corresponding issue](https://github.com/image-rs/image-tiff/issues/243) over at the `tiff` repository.

**Comment by weiji14** [comment]
> Hi, @weiji14! I managed to complete the migration to the `tiff` crate, although it might initially seem a bit radical. The corresponding PR is #17.

Nice, thanks for doing all that work! I was planning to do the migration more gradually in a series of PRs as part of #7, but reached a point where it probably made sense to do a complete rewrite. I'll try and give that PR a review this week.

> Next, I would do the coordinate transformation stuff, as, in my opinion, this is what should be provided by this crate on top of the functionality already in the `tiff` crate.

I've got some code to get the affine transform from a GeoTIFF that was implemented in https://github.com/weiji14/cog3pio/pull/8. Would be happy to upstream it to here afterwards.

**Comment by gschulze** [comment]
> I've got some code to get the affine transform from a GeoTIFF that was implemented in [weiji14/cog3pio#8](https://github.com/weiji14/cog3pio/pull/8). Would be happy to upstream it to here afterwards.

Hi, that sounds good! I already wrote some boilerplate code to handle the various combinations of `ModelTiepointTag`, `ModelPixelScaleTag` and `ModelTransformationTag`, according to what I understood from https://docs.ogc.org/is/19-008r4/19-008r4.html#_coordinate_transformations. You can find the current work-in-progress here: https://github.com/gschulze/geotiff/tree/feature/coordinate-transformation

There seem to be three cases for doing coordinate transformations:
1) Transformation by one tie point and pixel scale
2) Transformation by a transformation matrix
3) Transformation by using multiple tie points (or ground control points)

I think cases 1 and 2 are pretty straightforward; there are examples of how to do it properly, e.g., https://github.com/OSGeo/libgeotiff.

Case 3 is more complicated, and I did not find any standard way to do it. `libgeotiff` just has a `TODO` in the code at the corresponding location. One approach that comes to my mind would be to use Delaunay triangulation to decide the three control points that must be used for transforming a particular coordinate and then use linear interpolation to do the transformation.

I'm happy to continue with the implementation or leave it up to you if you want to. In that case, feel free to use anything you might find helpful from my feature branch.

By the way, things start to go a bit off-topic here. What do you think of opening a separate (tracking) issue for the things that need to be implemented so that we can better keep track of things?

**Comment by weiji14** [comment]
> I think cases 1 and 2 are pretty straightforward; there are examples of how to do it properly, e.g., https://github.com/OSGeo/libgeotiff.
> 
> Case 3 is more complicated, ...

Yep, I think I have a good handle on case 1 & 2, but 3 is an entirely different beast! We can leave a bunch of `unimplemented!()` calls for those in the short-term.

> By the way, things start to go a bit off-topic here. What do you think of opening a separate (tracking) issue for the things that need to be implemented so that we can better keep track of things?

Yes please! I'm usually hesitant to open (too many) issues when it's just one person doing all the work, preferring to just document things in PRs, but for >1 people, then it makes sense to have a proper checklist somewhere.

**Comment by feefladder** [comment]
I think such a complex thing as 3. would better be done by geodesy? They have an issue (busstoptaktik/geodesy#86), but I think that has a deformation for every pixel in our own tiff?

Anyways, the folks [over at QGIS](https://docs.qgis.org/3.34/en/docs/user_manual/working_with_raster/georeferencer.html#available-transformation-algorithms) use polynomial/spline transformations for their georeferencing.

**Comment by feefladder** [comment]
topic: COG
I just realized some time ago that an uncompressed cog is actually rather deterministic and even overviews can be written out-of-order from only the full-res image -once CRS and full image width is known.

![Image](https://github.com/user-attachments/assets/ba56e86e-4c72-4221-ad73-e98aaba55213)

but for compressed images (which is obviously better), the variable compression rates of different tiles (and only full tiles can be compressed) make that difficult.

**Comment by boxerab** [comment]
fwiw https://github.com/roteiro-gis/geotiff-rust has a pure Rust encoder with tiled writes, overviews and COG layout. As of roteiro-gis/geotiff-rust#91 the output passes GDAL's COG validator with full check on, and the writer runs on wasm32.

**Comment by feefladder** [comment]
O wow, that's seriously cool! Even with Lerc, deflate, jpeg.

I'm still slightly sad about the fact that COGs have the full-resolution image last, since it seriously hinders writing efficiency (the wasm has to keep full-res image in memory to the very end, since overviews cannot be determined). Is that actually used, or do all readers fetch tiles based on the offsets in the IFD data?

**Comment by kylebarron** [comment]
I'm not sure there's anything in the COG spec that _requires_ the full resolution _image data_ to be last in the file.

The **IFD** must be last, but I think that could point to a block of data immediately following the initial sequence of IFDs. ie for the image data to be in opposite order as the IFDs

**Comment by feefladder** [comment]
> I'm not sure there's anything in the COG spec that _requires_ the full resolution _image data_ to be last in the file.
>
> The **IFD** must be last, but I think that could point to a block of data immediately following the initial sequence of IFDs. ie for the image data to be in opposite order as the IFDs

Ah, you're right! It's a RECOMMENDATION! But so is IFD order then.. 
The relevant table is in [recommendation 3, paragraph 7.2.1](https://docs.ogc.org/is/21-026/21-026.html#_requirement_reduced_resolution_subfiles) points 1-8 for ifds and 9-12 for tile data. It does reccomend reverse IFD-data order, but exactly swapped: IFDs are specified in decreasing-resolution order e.g. linked-list-order-is-file-order-except-that-masks-are-interleaved-and-TileOffsets/Bytecounts-are-after-ifds (points 1-8*) and then the image data is laid out with linear streaming in mind, so low-res first.

I don't know any reader (GDAL, js geotiff or any in the Rust ecosystem) that _does_ linear streaming and if it's a recommendation only that's good news - a writer could write image data in decreasing-resolution order.

I had some more thoughts of possible optimizations for COG writing, that were somewhat blocked by tile-order ~~requirement~~ recommendation:
1. Have a single `NoData` tile, created at the start and all TileOffsets point to this tile until they are written. GDAL specially treats `TileOffsets==0` to mean "not written yet". With a single `NoData` tile, any tiles that do not contain data can be skipped completely and the file is always valid tiff. I think this is greatly beneficial, since all nodata is requested from the same region in the file, supercharging any cache wrt that.
2. Follow a space-filling curve (e.g. [Hilbert](https://github.com/becheran/fast-hilbert)) within overviews. I thought this would drastically reduce the number of tiles that need to be kept in memory, but the fact that overview `i` can't be written until after overview `i-1` messes that up. I derived upper bounds for the number of tiles in those cases. For an image of `n` by `n` tiles and half-size overviews, where `i` overviews don't need to be kept in-memory, that's `z=⌊log₂(n)⌋` overviews and ~`3(z-i)+1¼n²(¼)ⁱ` tiles for space-filling in stead of ~`2n(½)ⁱ+1¼n²(¼)ⁱ` tiles for row-major. For any reasonably-sized COGs, the `1¼n²(¼)ⁱ` vastly outweighs `3(z-i)` vs `2n(½)ⁱ`, so that's not really worth pursuing - unless the benefit of space-filling curves shows in reading.
3. Have the computation pipeline already write compressed, full-res tiles ~~in the right order~~. Then, the size of the full-res image is known in advance, so the first overview can also be written simultaneously. Based on 2., that reduces the memory footprint from `i=1` to `i=2`, or `~1/3 n²`-> `~1/12 n²` aka `¼`.

As for that, development of [tiff2](https://github.com/feefladder/tiff2) has also steadily progressed, but for the writing part, I got a bit lost in generifying "writing a tiff while allowing for COG writers". Now it's amazing that there's a working COG writer in Rust-on-wasm! I'll be looking at your code @boxerab for an-instance-to-be-generified if that's ok with you :)
